overlord/snapstate/snapstate.go

```diff
@@ -1216,42 +1216,63 @@ func resolveChannel(st *state.State, snapName, newChannel string, deviceCtx Devi
-			snap, strutil.ElliptRight(cohFrom, 10), strutil.ElliptRight(cohTo, 10))
-	case switchCohort && switchChannel && chanFrom == "" && cohFrom == "":
-		return fmt.Sprintf(i18n.G("Switch snap %q from no channel to %q and from no cohort to %q"),
-			snap, chanTo, strutil.ElliptRight(cohTo, 10),
-		)
-	case switchCohort && switchChannel && cohFrom == "":
-		return fmt.Sprintf(i18n.G("Switch snap %q from channel %q to %q and from no cohort to %q"),
-			snap, chanFrom, chanTo, strutil.ElliptRight(cohTo, 10),
-		)
-	case switchCohort && switchChannel && chanFrom == "":
-		return fmt.Sprintf(i18n.G("Switch snap %q from no channel to %q and from cohort %q to %q"),
-			snap, chanTo, strutil.ElliptRight(cohFrom, 10), strutil.ElliptRight(cohTo, 10),
-		)
-	case switchCohort && switchChannel:
+	if cohFrom != cohTo {
+		if cohTo == "" {
+			// leave cohort
+			if chanFrom == chanTo {
+				return fmt.Sprintf(i18n.G("Switch snap %q away from cohort %q"),
+					snap, strutil.ElliptRight(cohFrom, 10))
+			}
+			if chanFrom == "" {
+				return fmt.Sprintf(i18n.G("Switch snap %q from no channel to %q and away from cohort %q"),
+					snap, chanTo, strutil.ElliptRight(cohFrom, 10),
+				)
+			}
+			return fmt.Sprintf(i18n.G("Switch snap %q from channel %q to %q and away from cohort %q"),
+				snap, chanFrom, chanTo, strutil.ElliptRight(cohFrom, 10),
+			)
+		}
+		if cohFrom == "" {
+			// moving into a cohort
+			if chanFrom == chanTo {
+				return fmt.Sprintf(i18n.G("Switch snap %q from no cohort to %q"),
+					snap, strutil.ElliptRight(cohTo, 10))
+			}
+			if chanFrom == "" {
+				return fmt.Sprintf(i18n.G("Switch snap %q from no channel to %q and from no cohort to %q"),
+					snap, chanTo, strutil.ElliptRight(cohTo, 10),
+				)
+			}
+			// chanTo == "" is not interesting
+			return fmt.Sprintf(i18n.G("Switch snap %q from channel %q to %q and from no cohort to %q"),
+				snap, chanFrom, chanTo, strutil.ElliptRight(cohTo, 10),
+			)
+		}
+		if chanFrom == chanTo {
+			return fmt.Sprintf(i18n.G("Switch snap %q from cohort %q to %q"),
+				snap, strutil.ElliptRight(cohFrom, 10), strutil.ElliptRight(cohTo, 10))
+		}
+		if chanFrom == "" {
+			return fmt.Sprintf(i18n.G("Switch snap %q from no channel to %q and from cohort %q to %q"),
+				snap, chanTo, strutil.ElliptRight(cohFrom, 10), strutil.ElliptRight(cohTo, 10),
+			)
+		}
 		return fmt.Sprintf(i18n.G("Switch snap %q from channel %q to %q and from cohort %q to %q"),
 			snap, chanFrom, chanTo,
 			strutil.ElliptRight(cohFrom, 10), strutil.ElliptRight(cohTo, 10),
 		)
-	default:
-		// this might actually be an error
-		return "No change switch (bug?)"
 	}
+
+	if chanFrom == "" {
+		return fmt.Sprintf(i18n.G("Switch snap %q from no channel to %q"),
+			snap, chanTo)
+	}
+	if chanFrom != chanTo {
+		return fmt.Sprintf(i18n.G("Switch snap %q from channel %q to %q"),
+			snap, chanFrom, chanTo)
+	}
+	// this might actually be an error
+	return "No change switch (bug?)"
```

Comment:
I think this could use a better error message; should we simply say "nothing to do" (like we do on different occassion already somewhere else)?

Reply:
We need to decide if it's an error, and handle it as such -- or improve the message. Yes.

I'll do it in a followup (note the issue is not introduced in this PR)

Reply:
Further, the current snapd (before any cohort work) accepts a no-change switch and the change summary is `Switch snap "some-snap" to`. With this change it becomes `No change switch (bug?)` which I still think is better -- but it does need a decision on this.

Reply:
We don't want to error as we didn't so far and it seems reasonable for switch to be idempotent, we don't want to have "bug?" in a summary either though.  "No change switch (no-op)" seems the best compromise.

Reply:
I'm quite sure we don't want a "bug?" in a summary,

Reply:
as I tried to explain,  s/bug?/no-op/